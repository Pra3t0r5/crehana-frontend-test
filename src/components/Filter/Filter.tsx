import React, { FC, useEffect, useState } from "react";
import {
  Box,
  Checkbox,
  FormControlLabel,
  IconButton,
  ListItem,
  ListItemButton,
  ListItemIcon,
  ListItemText,
} from "@mui/material";
import "./Filter.css";
import { useCountriesQuery } from "../../generated/graphql";
import { DATASOURCE } from "../../constants";
import { CountryRowProps } from "../Country/Country";
import { StringifyOptions } from "querystring";

interface FilterProps {
  onFilter: (rows: CountryRowProps[]) => void;
  onLoading: (loading: boolean) => void;
  type: string;
}
interface CheckboxData {
  values: ICheck[];
}
interface ICheck {
  key: string;
  value: boolean;
}
interface FilteredProps {
  rows: (rows: CountryRowProps[]) => void;
  type: string;
}

// 1a creacion de atributos de checkboxes [{"AA":true}]
const generatedCheckData = (
  countries: any[] = [],
  type: string
): CheckboxData => {
  const DEFAULT_START = false;
  let keys: string[] = [];
  for (let country of countries) {
    let key;
    if (type === "currency" && !country.currency?.includes(null)) {
      key = country.currency;
    } else if (
      type === "continent" &&
      !country.continent?.code?.includes(null)
    ) {
      key = country.continent.code;
    }
    keys.push(key);
  }
  keys = keys.filter((value: string, index: number, array: string[]) => {
    return array.indexOf(value) === index;
  });
  const checkData = {
    // all: DEFAULT_START,
    values: keys.map((key: string) => ({ key, value: DEFAULT_START })),
  };
  console.log("generatedCheckData.checkData", checkData);
  return checkData;
};

const Filter: FC<FilterProps> = ({ onFilter, onLoading, type }) => {
  const [searchParams, setSearchParams] = useState<any>({});
  const { data: filterData, isFetching } = useCountriesQuery(
    DATASOURCE,
    searchParams,
    {
      queryKey: ["filterParams" + type, searchParams],
    }
  );

  const [checkData, setCheckData] = useState<CheckboxData>(
    generatedCheckData(filterData?.countries || [], type)
  );

  //2 seteo de checkboxdata como mapeo de busqueda
  const requestSearch = () => {
    const searchedArray = [];
    let filter = {};
    for (let chk of checkData.values) {
      if (chk.value) {
        searchedArray.push(chk.key);
      }
    }
    if (searchedArray.length > 0) {
      filter = {
        [type]: { in: searchedArray },
      };
    }
    setSearchParams({
      dataSource: DATASOURCE,
      filter,
    });
  };

  const handleChange = (event: any) => {
    const { checked, value: key } = event.target;
    const idx = checkData.values.findIndex(
      (check: ICheck) => check.key === key
    );
    if (idx !== -1) {
      setCheckData((prevState: any) => ({
        ...prevState,
        values: [
          ...prevState.values.slice(0, idx),
          { ...prevState.values[idx], value: checked },
          ...prevState.values.slice(idx + 1),
        ],
      }));
    }
  };

  //1b actualizacion de atributos de checkboxes
  useEffect(() => {
    if (checkData) {
      requestSearch();
      console.log("checkData.values", checkData.values);
    }
  }, [JSON.stringify(checkData.values)]);

  useEffect(() => {
    if (filterData?.countries) {
      onFilter(filterData?.countries || []);
    }
  }, [filterData?.countries]);

  useEffect(() => {
    onLoading(isFetching);
  }, [isFetching]);

  return (
    <Box
      className="Filter"
      data-testid="Filter"
      style={{ maxHeight: 600, overflow: "auto" }}
    >
      <>
        {/* <FormControlLabel
          label={type}
          control={
            <Checkbox checked={checkData.all} onChange={handleCheckAll} />
          }
        />*/}
        {checkData?.values?.map((obj: ICheck) => (
          <>
            <ListItem key={obj.key} disablePadding>
              <ListItemIcon>
                <Checkbox
                  edge="start"
                  checked={obj.value}
                  tabIndex={-1}
                  disableRipple
                  inputProps={{ "aria-labelledby": obj.key }}
                  value={obj.key}
                  onChange={handleChange}
                />
              </ListItemIcon>
              <ListItemText id={obj.key} primary={obj.key} />
            </ListItem>
          </>
        ))}
      </>
    </Box>
  );
};

export default Filter;
