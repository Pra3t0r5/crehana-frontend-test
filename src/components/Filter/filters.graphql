query Continents($filter: ContinentFilterInput = {}) {
  continents(filter: $filter) {
    code
    name
    countries {
      code
    }
  }
}
query Language($filter: LanguageFilterInput = {}) {
  languages(filter: $filter) {
    code
    name
    native
    rtl
  }
}
