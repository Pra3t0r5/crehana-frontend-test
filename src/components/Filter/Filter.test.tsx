import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Filter from "./Filter";
import { QueryClient, QueryClientProvider } from "react-query";
import renderer from 'react-test-renderer';

describe("<Filter />", () => {
  test("it should mount", () => {
    const queryClient = new QueryClient();
    const setRows = jest.fn();
    const setLoading = jest.fn();
    const type: string = "continent";
    const set = jest.fn();
    const searchParams = {};
    render(
      <QueryClientProvider contextSharing={true} client={queryClient}>
        <Filter
          onFilter={setRows}
          onLoading={setLoading}
          type={type}
          onSearchParamsChange={set}
          searchParams={searchParams}
        />
      </QueryClientProvider>
    );

    const filter = screen.getByTestId("Filter");

    expect(filter).toBeInTheDocument();
  });
  test('it should match snapshot', () => {
    const queryClient = new QueryClient();
    const setRows = jest.fn();
    const setLoading = jest.fn();
    const type: string = "continent";
    const set = jest.fn();
    const searchParams = {};
    const tree = renderer.create(
      <QueryClientProvider contextSharing={true} client={queryClient}>
        <Filter
          onFilter={setRows}
          onLoading={setLoading}
          type={type}
          onSearchParamsChange={set}
          searchParams={searchParams}
        />
      </QueryClientProvider>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
