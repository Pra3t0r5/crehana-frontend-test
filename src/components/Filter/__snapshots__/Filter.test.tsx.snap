// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filter /> it should match snapshot 1`] = `
<div
  className="MuiFormControl-root css-q8hpuo-MuiFormControl-root"
  data-testid="Filter"
>
  <label
    className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink={false}
  >
    continent
  </label>
  <div
    className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    onClick={[Function]}
    variant="outlined"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      className="MuiSelect-select MuiSelect-outlined MuiSelect-multiple MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        className="notranslate"
      >
        â€‹
      </span>
    </div>
    <input
      aria-hidden={true}
      className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
      disabled={false}
      onAnimationStart={[Function]}
      onChange={[Function]}
      required={false}
      tabIndex={-1}
      value=""
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
      data-testid="ArrowDropDownIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-1ftyaf0"
      >
        <span>
          continent
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
